designs/disruption-controls-by-method.md

```diff
@@ -0,0 +1,403 @@
+
+By separating method and reason, Karpenter allows users to define budgets and policies at both levels. Users can set a general budget for all "Drift" disruptions, and then further refine the control by specifying different budgets or schedules for specific reasons like "AMIDrift". This two-tiered approach offers flexibility and precision, enabling users to manage disruptions more effectively according to their needs.
+
+Moreover, the distinction helps in maintaining clarity and organization within the API. It allows for a structured way to handle disruptions, where methods can be seen as categories, and reasons as subcategories. This hierarchy makes it easier for users to navigate and understand the disruption policies they have set up.
+
+This also allows karpenter to easily tell inside of the disruption controller which actions it needs to be looking for when checking Type(). Without this top method, it becomes much more challenging to match a Type() with a Reason. 
+
+#### Q: Budgets currently work by tracking deletion in total. In this new system, karpenter has to be aware of each disruption method + reason, does adding two fields Method + Reason make it harder to track how much of budget we have used? Should we add DisruptionReason to the nodeclaim? 
+To properly track nodeclaims in deleting state for each nodepool effectively and easily in cluster state adding an additional field or status condition to indicate why the nodeclaim is being disrupted/deleted  makes a lot of sense. A single reason makes it easier to track on the nodeclaim.
+##### Q: Should we have two status conditions for disruption method and disruption reason? Or should they be consolidated into one condition?
+Having two separate status conditions for disruption method and disruption reason could potentially provide more detailed information about the disruption. However, it might also complicate the tracking process. On the other hand, consolidating them into one condition would simplify the tracking but might lack some details. Considering the trade-off between detail and simplicity, it would be more practical to consolidate them into one condition. This way, we can track the disruption process more efficiently while still maintaining necessary information about the disruption.
+
+```go
+type NodeClaimStatus struct {
+	...
+	// DisruptionDetails represents the method and reason for the disruption
+	// in the format of "method:reason".
+	DisruptionDetails string `json:"disruptionDetails,omitempty"`
+	...
+}
+```
+#### Q: For a conflicting child reason, do we respect the parent method? 
+```yaml 
+spec:
+  disruption:
+    budgets:
+    - nodes: 50
+      method: "Drift" 
+      reason: "NodeImageDrift"
```

Comment:
Q: Should we add validation for reasons? If so how do we do it with CEL?   
Reasons are a bit complex to validate since some of them can be defined per cloud provider. So far in Karpenter, when there is a cloud provider specific concept, it goes in the NodeClass. In the case of drift however, cloudproviders return a drift Reason back to core.

However, if you reflect on instance type requirements cloudproviders will let karpenter know their WellKnownLabels. We could have a similar registration of acceptable drift reasons. The problem here is that we are using CEL validation, and moving away from webhooks. This type of validation requires our validation to have access with this type of map.

Alternatively we could just hardcode Provider Drift Reasons into the core validation for reason, but then you define clear coupling for all drift reasons between core and the providers. That is also undesirable

If we have no validation, users may specify nonsensical values for budgets, and since karpenter across the board has estabilished a user experience where we reject these nonsensical values, users will expect that same experience here.

So, what do we do here? Thoughts @garvinp-stripe?

Reply:
I unfortunately am not very in tune with how we want to deal with the provider level validation but you are right this raises a important question on how should we do this. Maybe @jonathan-innis has some thoughts